### PR TITLE
Enhancements for local proto behavior

### PR DESCRIPTION
**Pull request by ad-daniel**
**Description**

Improvements to the EXTERNPROTO  logic for locally available files:

- [x] fix backwards compatibility logic applying too early
- [x] show warning when a non-declared PROTO is known to be available
- [x] on save, write EXTERNPROTO exactly as it was loaded (currently: relative path is saved as absolute)
- [x] show warning in ambiguous cases

Additionally, it fixes: https://github.com/cyberbotics/webots/issues/4720
Related discussion: https://github.com/cyberbotics/webots/pull/4751
- [x] gray out clashing proto
- [x] fix addnode showing hashes

### CODE REVIEW

**Comment by ad-daniel** [comment]
Note: proto clash avoidance during insertion is addressed in this PR: https://github.com/cyberbotics/webots/pull/4751

File: src/webots/vrml/WbProtoManager.hpp [hunk 1]
```diff
@@ -113,14 +117,14 @@ class WbProtoManager : public QObject {
   enum { BASE_NODE = 10001, PROTO_WORLD = 10002, PROTO_PROJECT = 10003, PROTO_EXTRA = 10004, PROTO_WEBOTS = 10005 };
 
   // given a category and a PROTO name, it returns the url from the corresponding list
-  QString protoUrl(int category, const QString &protoName) const;
+  QString protoUrl(const QString &protoName, int category);
```
Comment (omichel) [inline src/webots/vrml/WbProtoManager.hpp:120]: Why did you remove the const qualifier? It seems this method doesn't modify the object:
```suggestion
  QString protoUrl(const QString &protoName, int category) const;
```

File: src/webots/scene_tree/WbAddNodeDialog.cpp [hunk 2]
```diff
@@ -673,6 +685,30 @@ void WbAddNodeDialog::import() {
...
+  if (WbUrl::isLocalUrl(thisUrl))
+    thisUrl = QDir::cleanPath(WbStandardPaths::webotsHomePath() + url.mid(9));
+  else
+    thisUrl = mUniqueLocalProto.value(protoName);
+
+  QString otherUrl;
+  if (WbUrl::isLocalUrl(thisUrl))
+    thisUrl = QDir::cleanPath(WbStandardPaths::webotsHomePath() + url.mid(9));
+  else
+    thisUrl = url;
```
Comment (omichel) [inline src/webots/scene_tree/WbAddNodeDialog.cpp:704]: Seems wrong (didn't check clang-format):
```suggestion
  const QString thisUrl = WbUrl::isLocalUrl(url) ? QDir::cleanPath(WbStandardPaths::webotsHomePath() + url.mid(9)) : mUniqueLocalProto.value(protoName);

  const QString otherUrl = WbUrl::isLocalUrl(url) ? QDir::cleanPath(WbStandardPaths::webotsHomePath() + url.mid(9)) : url;
```

File: src/webots/vrml/WbProtoTreeItem.hpp [hunk 3]
```diff
@@ -30,6 +30,9 @@ class WbProtoTreeItem : public QObject {
   const WbProtoTreeItem *parent() const { return mParent; }
   const QList<WbProtoTreeItem *> children() const { return mChildren; }
 
+  void setRawUrl(const QString &url) { mRawUrl = url; }
+  const QString &rawUrl() const { return mRawUrl; }
```
Comment (omichel) [inline src/webots/vrml/WbProtoTreeItem.hpp:34]: It seem this method is never called, and the value of `mRawUrl` is also never used.

**Comment by ad-daniel** [comment]
I believe I've addressed all issues, so the PR can be reviewed while waiting for the test results

File: src/webots/scene_tree/WbAddNodeDialog.cpp [hunk 4]
```diff
@@ -687,22 +687,19 @@ void WbAddNodeDialog::import() {
...
+    thisUrl = QDir::cleanPath(WbStandardPaths::webotsHomePath() + thisUrl.mid(9));
 
-  QString otherUrl;
-  if (WbUrl::isLocalUrl(thisUrl))
-    thisUrl = QDir::cleanPath(WbStandardPaths::webotsHomePath() + url.mid(9));
-  else
-    thisUrl = url;
+  QString otherUrl = url;
+  if (WbUrl::isLocalUrl(otherUrl))
+    otherUrl = QDir::cleanPath(WbStandardPaths::webotsHomePath() + otherUrl.mid(9));
```
[the remark this answers is not in the document]
Comment (ad-daniel) [reply src/webots/scene_tree/WbAddNodeDialog.cpp:702]: should be `url.mid(9)`, not `otherUrl.mid(9)` since it is still undefined, no?

[the remark this answers is not in the document]
Comment (omichel) [reply src/webots/scene_tree/WbAddNodeDialog.cpp]: Yes, sorry I missed that.

[the remark this answers is not in the document]
Comment (omichel) [reply src/webots/scene_tree/WbAddNodeDialog.cpp]: I just fixed my suggestion.

**Review by omichel — APPROVED** [review]
Thank you.
